src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp

```diff
@@ -2954,6 +2954,92 @@ void C2_MacroAssembler::reduce_integral_v(Register dst, Register src1,
   vmv_x_s(dst, tmp);
 }
 
+void C2_MacroAssembler::reduce_mul_integer_v(Register dst, Register src1, VectorRegister src2,
+                                             VectorRegister vtmp1, VectorRegister vtmp2,
+                                             BasicType bt, uint vector_length, VectorMask vm) {
+  assert(bt == T_BYTE || bt == T_SHORT || bt == T_INT || bt == T_LONG, "unsupported element type");
+  uint len = vector_length/type2aelembytes(bt);
+  vsetvli_helper(bt, len);
+
+  len /= 2;
+  if (vm != Assembler::unmasked) {
+    // This behaviour is consistent with spec requirements of vector API, for `reduceLanes`:
+    //  If no elements are selected, an operation-specific identity value is returned.
+    //    If the operation is MUL, then the identity value is one.
+    vmv_v_i(vtmp1, 1);
+    vmerge_vvm(vtmp2, vtmp1, src2); // vm == v0
+    vslidedown_vi(vtmp1, vtmp2, len);
+
+    vsetvli_helper(bt, len);
+    vmul_vv(vtmp1, vtmp1, vtmp2);
+  } else {
+    vslidedown_vi(vtmp1, src2, len);
+
+    vsetvli_helper(bt, len);
+    vmul_vv(vtmp1, vtmp1, src2);
+  }
+
+  while (len > 1) {
+    len /= 2;
+    vslidedown_vi(vtmp2, vtmp1, len);
+    vsetvli_helper(bt, len);
+    vmul_vv(vtmp1, vtmp1, vtmp2);
+  }
+
+  vmv_x_s(dst, vtmp1);
+  if (bt == T_INT) {
+    mulw(dst, dst, src1);
+  } else {
+    mul(dst, dst, src1);
+  }
+}
+
+void C2_MacroAssembler::reduce_mul_fp_v(FloatRegister dst, FloatRegister src1, VectorRegister src2,
+                                        VectorRegister vtmp1, VectorRegister vtmp2,
+                                        BasicType bt, uint vector_length, VectorMask vm) {
+  assert(bt == T_FLOAT || bt == T_DOUBLE, "unsupported element type");
+  uint len = vector_length/type2aelembytes(bt);
+  vsetvli_helper(bt, len);
+
+  len /= 2;
+  if (vm != Assembler::unmasked) {
+    // This behaviour is consistent with spec requirements of vector API, for `reduceLanes`:
+    //  If no elements are selected, an operation-specific identity value is returned.
+    //    If the operation is MUL, then the identity value is one.
+    if (bt == T_FLOAT) {
+      lui(t0, 0x3f800000); // 1.0f
+    } else {
+      lui(t0, 0x3ff00000); // 1.0d
```

Comment:
Can you use `mv` instead of `lui`, like other places?

Reply:
fixed.